File: swift_reference/Services/APIService.swift
import Foundation
import AuthenticationServices

nonisolated struct APIResponse<T: Decodable & Sendable>: Decodable, Sendable {
    let result: APIResult<T>
}

nonisolated struct APIResult<T: Decodable & Sendable>: Decodable, Sendable {
    let data: T
}

nonisolated struct AuthResponse: Codable, Sendable {
    let userId: String
    let displayName: String?
    let isNewUser: Bool
}

nonisolated struct CircleResponse: Codable, Sendable {
    let id: String
    let name: String
    let inviteCode: String
}

nonisolated struct CircleListItem: Codable, Sendable, Identifiable {
    let id: String
    let name: String
    let description: String
    let memberCount: Int
    let role: String
    let inviteCode: String
}

nonisolated struct CircleDetail: Codable, Sendable {
    let id: String
    let name: String
    let description: String
    let memberCount: Int
    let inviteCode: String
    let createdAt: String
    let members: [CircleMemberInfo]
}

nonisolated struct CircleMemberInfo: Codable, Sendable, Identifiable {
    let userId: String
    let role: String
    let joinedAt: String

    var id: String { userId }
}

nonisolated struct JoinCircleResponse: Codable, Sendable {
    let id: String
    let name: String
    let alreadyMember: Bool
}

nonisolated struct SOSSendResponse: Codable, Sendable {
    let id: String
    let recipientCount: Int
}

nonisolated struct SOSItem: Codable, Sendable, Identifiable {
    let id: String
    let senderId: String
    let circleId: String
    let message: String
    let createdAt: String
    let isMine: Bool
}

nonisolated struct ShareLinkResponse: Codable, Sendable {
    let shareUrl: String
    let inviteCode: String
}

nonisolated struct SundaySummaryResponse: Codable, Sendable {
    let circleId: String
    let weekOf: String
    let totalMembers: Int
    let activeMembers: Int
    let averageScore: Double
    let topStreaks: [TopStreak]
}

nonisolated struct TopStreak: Codable, Sendable, Identifiable {
    let userId: String
    let streak: Int
    var id: String { userId }
}

nonisolated struct CircleHeatmapResponse: Codable, Sendable {
    let circleId: String
    let aggregatedData: [CircleHeatmapDay]
}

nonisolated struct CircleHeatmapDay: Codable, Sendable, Identifiable {
    let date: String
    let averageScore: Double
    let memberCount: Int
    var id: String { date }
}

nonisolated struct SOSContactsResponse: Codable, Sendable {
    let circleId: String
    let contactCount: Int
}

nonisolated struct SharedGratitudeItem: Codable, Sendable, Identifiable {
    let id: String
    let gratitudeText: String
    let isAnonymous: Bool
    let displayName: String?
    let sharedAt: String
    let isMine: Bool
}

nonisolated struct GratitudeWallResponse: Codable, Sendable {
    let circleId: String
    let weeksBack: Int
    let gratitudes: [SharedGratitudeItem]
}

nonisolated struct ShareGratitudeResponse: Codable, Sendable {
    let shared: [ShareGratitudeResult]
}

nonisolated struct ShareGratitudeResult: Codable, Sendable {
    let circleId: String
    let gratitudeId: String
}

nonisolated struct GratitudeNewCountResponse: Codable, Sendable {
    let circleId: String
    let newCount: Int
}

nonisolated struct GratitudeWeekCountResponse: Codable, Sendable {
    let circleId: String
    let weekCount: Int
}

@Observable
@MainActor
class APIService {
    static let shared = APIService()

    private(set) var userId: String?
    private(set) var isAuthenticated: Bool = false

    private let session = URLSession.shared
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }()

    private var baseURL: String {
        Config.EXPO_PUBLIC_RORK_API_BASE_URL
    }

    init() {
        userId = UserDefaults.standard.string(forKey: "tribute_user_id")
        isAuthenticated = userId != nil
    }

    func signInWithApple(identityToken: String, authorizationCode: String, fullName: PersonNameComponents?, email: String?) async throws -> AuthResponse {
        var nameDict: [String: String?]? = nil
        if let fullName {
            nameDict = [
                "givenName": fullName.givenName,
                "familyName": fullName.familyName,
            ]
        }

        let body: [String: Any?] = [
            "identityToken": identityToken,
            "authorizationCode": authorizationCode,
            "fullName": nameDict as Any,
            "email": email,
        ]

        let result: AuthResponse = try await postMutation("auth.signInWithApple", body: body)

        userId = result.userId
        isAuthenticated = true
        UserDefaults.standard.set(result.userId, forKey: "tribute_user_id")

        return result
    }

    func signOut() {
        userId = nil
        isAuthenticated = false
        UserDefaults.standard.removeObject(forKey: "tribute_user_id")
    }

    func createCircle(name: String, description: String = "") async throws -> CircleResponse {
        let body: [String: Any] = ["name": name, "description": description]
        return try await postMutation("circles.create", body: body)
    }

    func joinCircle(inviteCode: String) async throws -> JoinCircleResponse {
        let body: [String: Any] = ["inviteCode": inviteCode]
        return try await postMutation("circles.join", body: body)
    }

    func listCircles() async throws -> [CircleListItem] {
        return try await getQuery("circles.list")
    }

    func getCircleDetail(circleId: String) async throws -> CircleDetail {
        let input: [String: Any] = ["circleId": circleId]
        return try await getQuery("circles.getDetail", input: input)
    }

    func leaveCircle(circleId: String) async throws {
        let body: [String: Any] = ["circleId": circleId]
        let _: [String: Bool] = try await postMutation("circles.leave", body: body)
    }

    func sendSOS(circleId: String, message: String, recipientIds: [String]) async throws -> SOSSendResponse {
        let body: [String: Any] = [
            "circleId": circleId,
            "message": message,
            "recipientIds": recipientIds,
        ]
        return try await postMutation("sos.send", body: body)
    }

    func getRecentSOS(circleId: String? = nil, limit: Int = 20) async throws -> [SOSItem] {
        var input: [String: Any] = ["limit": limit]
        if let circleId { input["circleId"] = circleId }
        return try await getQuery("sos.getRecent", input: input)
    }

    func submitHeatmapData(circleId: String, weekData: [[String: Any]]) async throws {
        let body: [String: Any] = ["circleId": circleId, "weekData": weekData]
        let _: [String: Bool] = try await postMutation("circles.submitHeatmapData", body: body)
    }

    func generateShareLink(circleId: String) async throws -> ShareLinkResponse {
        let body: [String: Any] = ["circleId": circleId]
        return try await postMutation("invite.generateShareLink", body: body)
    }

    func getSundaySummary(circleId: String) async throws -> SundaySummaryResponse {
        let input: [String: Any] = ["circleId": circleId]
        return try await getQuery("circles.getSundaySummary", input: input)
    }

    func getCircleHeatmap(circleId: String, weeksBack: Int = 12) async throws -> CircleHeatmapResponse {
        let input: [String: Any] = ["circleId": circleId, "weeksBack": weeksBack]
        return try await getQuery("circles.getCircleHeatmap", input: input)
    }

    func setSOSContacts(circleId: String, contactUserIds: [String]) async throws -> SOSContactsResponse {
        let body: [String: Any] = ["circleId": circleId, "contactUserIds": contactUserIds]
        return try await postMutation("sos.setSOSContacts", body: body)
    }

    func registerPushToken(_ token: String) async {
        guard let userId else { return }
        let body: [String: Any] = ["userId": userId, "pushToken": token]
        do {
            let _: [String: Bool] = try await postMutation("auth.registerPushToken", body: body)
        } catch {
        }
    }

    func shareGratitude(circleIds: [String], gratitudeText: String, isAnonymous: Bool, displayName: String?) async throws -> ShareGratitudeResponse {
        var body: [String: Any] = [
            "circleIds": circleIds,
            "gratitudeText": gratitudeText,
            "isAnonymous": isAnonymous,
        ]
        if let displayName {
            body["displayName"] = displayName
        }
        return try await postMutation("gratitudes.share", body: body)
    }

    func getGratitudeWall(circleId: String, weeksBack: Int = 0) async throws -> GratitudeWallResponse {
        let input: [String: Any] = ["circleId": circleId, "weeksBack": weeksBack]
        return try await getQuery("gratitudes.getWall", input: input)
    }

    func deleteGratitude(circleId: String, gratitudeId: String) async throws {
        let body: [String: Any] = ["circleId": circleId, "gratitudeId": gratitudeId]
        let _: [String: Bool] = try await postMutation("gratitudes.delete", body: body)
    }

    func getGratitudeNewCount(circleId: String) async throws -> GratitudeNewCountResponse {
        let input: [String: Any] = ["circleId": circleId]
        return try await getQuery("gratitudes.getNewCount", input: input)
    }

    func markGratitudesSeen(circleId: String) async throws {
        let body: [String: Any] = ["circleId": circleId]
        let _: [String: Bool] = try await postMutation("gratitudes.markSeen", body: body)
    }

    func getGratitudeWeekCount(circleId: String) async throws -> GratitudeWeekCountResponse {
        let input: [String: Any] = ["circleId": circleId]
        return try await getQuery("gratitudes.getWeekCount", input: input)
    }

    private func getQuery<T: Decodable & Sendable>(_ procedure: String, input: [String: Any]? = nil) async throws -> T {
        var urlString = "\(baseURL)/api/trpc/\(procedure)"

        if let input {
            let jsonData = try JSONSerialization.data(withJSONObject: input)
            let inputString = String(data: jsonData, encoding: .utf8) ?? "{}"
            let encoded = inputString.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
            urlString += "?input=\(encoded)"
        }

        guard let url = URL(string: urlString) else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        addAuthHeaders(&request)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw APIError.serverError
        }

        let decoded = try decoder.decode(APIResponse<T>.self, from: data)
        return decoded.result.data
    }

    private func postMutation<T: Decodable & Sendable>(_ procedure: String, body: [String: Any?]) async throws -> T {
        guard let url = URL(string: "\(baseURL)/api/trpc/\(procedure)") else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        addAuthHeaders(&request)

        let jsonData = try JSONSerialization.data(withJSONObject: body)
        request.httpBody = jsonData

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw APIError.serverError
        }

        let decoded = try decoder.decode(APIResponse<T>.self, from: data)
        return decoded.result.data
    }

    private func addAuthHeaders(_ request: inout URLRequest) {
        if let userId {
            request.setValue(userId, forHTTPHeaderField: "x-user-id")
        }
    }
}

nonisolated enum APIError: Error, LocalizedError, Sendable {
    case invalidURL
    case serverError
    case unauthorized
    case decodingError

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid request URL"
        case .serverError: return "Server error. Please try again."
        case .unauthorized: return "Please sign in to continue"
        case .decodingError: return "Unexpected response format"
        }
    }
}

File: lib/data/datasources/remote/api_service.dart
import 'dart:convert';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:http/http.dart' as http;

// Response models

class AuthResponse {
  final String userId;
  final String? displayName;
  final bool isNewUser;
  AuthResponse({required this.userId, this.displayName, required this.isNewUser});
  factory AuthResponse.fromJson(Map<String, dynamic> j) => AuthResponse(
    userId: j['userId'] as String,
    displayName: j['displayName'] as String?,
    isNewUser: j['isNewUser'] as bool? ?? false,
  );
}

class CircleResponse {
  final String id;
  final String name;
  final String inviteCode;
  CircleResponse({required this.id, required this.name, required this.inviteCode});
  factory CircleResponse.fromJson(Map<String, dynamic> j) => CircleResponse(
    id: j['id'] as String, name: j['name'] as String, inviteCode: j['inviteCode'] as String,
  );
}

class CircleListItem {
  final String id;
  final String name;
  final String description;
  final int memberCount;
  final String role;
  final String inviteCode;
  CircleListItem({required this.id, required this.name, required this.description, required this.memberCount, required this.role, required this.inviteCode});
  factory CircleListItem.fromJson(Map<String, dynamic> j) => CircleListItem(
    id: j['id'] as String, name: j['name'] as String, description: j['description'] as String? ?? '',
    memberCount: j['memberCount'] as int? ?? 0, role: j['role'] as String? ?? 'member',
    inviteCode: j['inviteCode'] as String? ?? '',
  );
}

class CircleDetail {
  final String id;
  final String name;
  final String description;
  final int memberCount;
  final String inviteCode;
  final String createdAt;
  final List<CircleMemberInfo> members;
  CircleDetail({required this.id, required this.name, required this.description, required this.memberCount, required this.inviteCode, required this.createdAt, required this.members});
  factory CircleDetail.fromJson(Map<String, dynamic> j) => CircleDetail(
    id: j['id'] as String, name: j['name'] as String, description: j['description'] as String? ?? '',
    memberCount: j['memberCount'] as int? ?? 0, inviteCode: j['inviteCode'] as String? ?? '',
    createdAt: j['createdAt'] as String? ?? '',
    members: (j['members'] as List<dynamic>? ?? []).map((m) => CircleMemberInfo.fromJson(m as Map<String, dynamic>)).toList(),
  );
}

class CircleMemberInfo {
  final String userId;
  final String role;
  final String joinedAt;
  CircleMemberInfo({required this.userId, required this.role, required this.joinedAt});
  factory CircleMemberInfo.fromJson(Map<String, dynamic> j) => CircleMemberInfo(
    userId: j['userId'] as String, role: j['role'] as String? ?? 'member', joinedAt: j['joinedAt'] as String? ?? '',
  );
}

class JoinCircleResponse {
  final String id;
  final String name;
  final bool alreadyMember;
  JoinCircleResponse({required this.id, required this.name, required this.alreadyMember});
  factory JoinCircleResponse.fromJson(Map<String, dynamic> j) => JoinCircleResponse(
    id: j['id'] as String, name: j['name'] as String, alreadyMember: j['alreadyMember'] as bool? ?? false,
  );
}

class SOSSendResponse {
  final String id;
  final int recipientCount;
  SOSSendResponse({required this.id, required this.recipientCount});
  factory SOSSendResponse.fromJson(Map<String, dynamic> j) => SOSSendResponse(
    id: j['id'] as String, recipientCount: j['recipientCount'] as int? ?? 0,
  );
}

class SOSItem {
  final String id;
  final String senderId;
  final String circleId;
  final String message;
  final String createdAt;
  final bool isMine;
  SOSItem({required this.id, required this.senderId, required this.circleId, required this.message, required this.createdAt, required this.isMine});
  factory SOSItem.fromJson(Map<String, dynamic> j) => SOSItem(
    id: j['id'] as String, senderId: j['senderId'] as String, circleId: j['circleId'] as String,
    message: j['message'] as String, createdAt: j['createdAt'] as String, isMine: j['isMine'] as bool? ?? false,
  );
}

class ShareLinkResponse {
  final String shareUrl;
  final String inviteCode;
  ShareLinkResponse({required this.shareUrl, required this.inviteCode});
  factory ShareLinkResponse.fromJson(Map<String, dynamic> j) => ShareLinkResponse(
    shareUrl: j['shareUrl'] as String, inviteCode: j['inviteCode'] as String,
  );
}

class SundaySummaryResponse {
  final String circleId;
  final String weekOf;
  final int totalMembers;
  final int activeMembers;
  final double averageScore;
  final List<TopStreak> topStreaks;
  SundaySummaryResponse({required this.circleId, required this.weekOf, required this.totalMembers, required this.activeMembers, required this.averageScore, required this.topStreaks});
  factory SundaySummaryResponse.fromJson(Map<String, dynamic> j) => SundaySummaryResponse(
    circleId: j['circleId'] as String, weekOf: j['weekOf'] as String,
    totalMembers: j['totalMembers'] as int? ?? 0, activeMembers: j['activeMembers'] as int? ?? 0,
    averageScore: (j['averageScore'] as num?)?.toDouble() ?? 0,
    topStreaks: (j['topStreaks'] as List<dynamic>? ?? []).map((s) => TopStreak.fromJson(s as Map<String, dynamic>)).toList(),
  );
}

class TopStreak {
  final String userId;
  final int streak;
  TopStreak({required this.userId, required this.streak});
  factory TopStreak.fromJson(Map<String, dynamic> j) => TopStreak(userId: j['userId'] as String, streak: j['streak'] as int? ?? 0);
}

class SharedGratitudeItem {
  final String id;
  final String gratitudeText;
  final bool isAnonymous;
  final String? displayName;
  final String sharedAt;
  final bool isMine;
  SharedGratitudeItem({required this.id, required this.gratitudeText, required this.isAnonymous, this.displayName, required this.sharedAt, required this.isMine});
  factory SharedGratitudeItem.fromJson(Map<String, dynamic> j) => SharedGratitudeItem(
    id: j['id'] as String, gratitudeText: j['gratitudeText'] as String,
    isAnonymous: j['isAnonymous'] as bool? ?? false, displayName: j['displayName'] as String?,
    sharedAt: j['sharedAt'] as String, isMine: j['isMine'] as bool? ?? false,
  );
}

class GratitudeWallResponse {
  final String circleId;
  final int weeksBack;
  final List<SharedGratitudeItem> gratitudes;
  GratitudeWallResponse({required this.circleId, required this.weeksBack, required this.gratitudes});
  factory GratitudeWallResponse.fromJson(Map<String, dynamic> j) => GratitudeWallResponse(
    circleId: j['circleId'] as String, weeksBack: j['weeksBack'] as int? ?? 0,
    gratitudes: (j['gratitudes'] as List<dynamic>? ?? []).map((g) => SharedGratitudeItem.fromJson(g as Map<String, dynamic>)).toList(),
  );
}

class GratitudeNewCountResponse {
  final String circleId;
  final int newCount;
  GratitudeNewCountResponse({required this.circleId, required this.newCount});
  factory GratitudeNewCountResponse.fromJson(Map<String, dynamic> j) => GratitudeNewCountResponse(
    circleId: j['circleId'] as String, newCount: j['newCount'] as int? ?? 0,
  );
}

class GratitudeWeekCountResponse {
  final String circleId;
  final int weekCount;
  GratitudeWeekCountResponse({required this.circleId, required this.weekCount});
  factory GratitudeWeekCountResponse.fromJson(Map<String, dynamic> j) => GratitudeWeekCountResponse(
    circleId: j['circleId'] as String, weekCount: j['weekCount'] as int? ?? 0,
  );
}

class SOSContactsResponse {
  final String circleId;
  final int contactCount;
  SOSContactsResponse({required this.circleId, required this.contactCount});
  factory SOSContactsResponse.fromJson(Map<String, dynamic> j) => SOSContactsResponse(
    circleId: j['circleId'] as String, contactCount: j['contactCount'] as int? ?? 0,
  );
}

class CircleHeatmapDay {
  final String date;
  final double intensity;
  CircleHeatmapDay({required this.date, required this.intensity});
  factory CircleHeatmapDay.fromJson(Map<String, dynamic> j) => CircleHeatmapDay(
    date: j['date'] as String,
    intensity: (j['intensity'] as num?)?.toDouble() ?? 0.0,
  );
}

class CircleHeatmapResponse {
  final String circleId;
  final int weekCount;
  final List<CircleHeatmapDay> days;
  CircleHeatmapResponse({required this.circleId, required this.weekCount, required this.days});
  factory CircleHeatmapResponse.fromJson(Map<String, dynamic> j) => CircleHeatmapResponse(
    circleId: j['circleId'] as String,
    weekCount: j['weekCount'] as int? ?? 1,
    days: (j['days'] as List<dynamic>? ?? []).map((d) => CircleHeatmapDay.fromJson(d as Map<String, dynamic>)).toList(),
  );
}

class CircleMilestone {
  final String id;
  final String title;
  final String message;
  final String achievedAt;
  CircleMilestone({required this.id, required this.title, required this.message, required this.achievedAt});
  factory CircleMilestone.fromJson(Map<String, dynamic> j) => CircleMilestone(
    id: j['id'] as String,
    title: j['title'] as String,
    message: j['message'] as String,
    achievedAt: j['achievedAt'] as String,
  );
}

class CircleMilestonesResponse {
  final String circleId;
  final int totalGivingDays;
  final double totalHours;
  final int totalGratitudeDays;
  final List<CircleMilestone> milestones;
  CircleMilestonesResponse({required this.circleId, required this.totalGivingDays, required this.totalHours, required this.totalGratitudeDays, required this.milestones});
  factory CircleMilestonesResponse.fromJson(Map<String, dynamic> j) => CircleMilestonesResponse(
    circleId: j['circleId'] as String,
    totalGivingDays: j['totalGivingDays'] as int? ?? 0,
    totalHours: (j['totalHours'] as num?)?.toDouble() ?? 0.0,
    totalGratitudeDays: j['totalGratitudeDays'] as int? ?? 0,
    milestones: (j['milestones'] as List<dynamic>? ?? []).map((m) => CircleMilestone.fromJson(m as Map<String, dynamic>)).toList(),
  );
}

// Errors

class APIError implements Exception {
  final String message;
  const APIError._(this.message);

  static const invalidURL = APIError._('Invalid request URL');
  static const serverError = APIError._('Server error. Please try again.');
  static const unauthorized = APIError._('Please sign in to continue');
  static const decodingError = APIError._('Unexpected response format');

  @override
  String toString() => message;
}

// Service

class APIService {
  static final APIService shared = APIService._();
  APIService._();

  String? userId;
  bool isAuthenticated = false;

  static const String _baseURL = String.fromEnvironment(
    'API_BASE_URL',
    defaultValue: 'https://your-backend-url.com',
  );

  Future<void> init() async {
    final user = FirebaseAuth.instance.currentUser;
    if (user != null) {
      userId = user.uid;
      isAuthenticated = true;
    }
  }

  void setFirebaseToken(String? token, {String? userId}) {
    this.userId = userId ?? FirebaseAuth.instance.currentUser?.uid;
    isAuthenticated = token != null;
  }

  Future<AuthResponse> ensureProfile({String? displayName, String? email}) {
    final body = <String, dynamic>{
      if (displayName != null) 'displayName': displayName,
      if (email != null) 'email': email,
    };
    return _postMutation<AuthResponse>(
      'auth.ensureProfile',
      body: body,
      fromJson: (j) => AuthResponse.fromJson(j),
    );
  }

  Future<void> signOut() async {
    userId = null;
    isAuthenticated = false;
  }

  Future<CircleResponse> createCircle(String name, {String description = ''}) =>
      _postMutation('circles.create', body: {'name': name, 'description': description}, fromJson: (j) => CircleResponse.fromJson(j));

  Future<JoinCircleResponse> joinCircle(String inviteCode) =>
      _postMutation('circles.join', body: {'inviteCode': inviteCode}, fromJson: (j) => JoinCircleResponse.fromJson(j));

  Future<List<CircleListItem>> listCircles() =>
      _getQuery('circles.list', fromJson: (j) => (j as List<dynamic>).map((e) => CircleListItem.fromJson(e as Map<String, dynamic>)).toList());

  Future<CircleDetail> getCircleDetail(String circleId) =>
      _getQuery('circles.getDetail', input: {'circleId': circleId}, fromJson: (j) => CircleDetail.fromJson(j));

  Future<void> leaveCircle(String circleId) =>
      _postMutation<Map<String, dynamic>>('circles.leave', body: {'circleId': circleId}, fromJson: (j) => j);

  Future<SOSSendResponse> sendSOS(String circleId, String message, List<String> recipientIds) =>
      _postMutation('sos.send', body: {'circleId': circleId, 'message': message, 'recipientIds': recipientIds}, fromJson: (j) => SOSSendResponse.fromJson(j));

  Future<List<SOSItem>> getRecentSOS({String? circleId, int limit = 20}) =>
      _getQuery('sos.getRecent', input: {'limit': limit, if (circleId != null) 'circleId': circleId}, fromJson: (j) => (j as List<dynamic>).map((e) => SOSItem.fromJson(e as Map<String, dynamic>)).toList());

  Future<ShareLinkResponse> generateShareLink(String circleId) =>
      _postMutation('invite.generateShareLink', body: {'circleId': circleId}, fromJson: (j) => ShareLinkResponse.fromJson(j));

  Future<SundaySummaryResponse> getSundaySummary(String circleId) =>
      _getQuery('circles.getSundaySummary', input: {'circleId': circleId}, fromJson: (j) => SundaySummaryResponse.fromJson(j));

  Future<SOSContactsResponse> setSOSContacts(String circleId, List<String> contactUserIds) =>
      _postMutation('sos.setSOSContacts', body: {'circleId': circleId, 'contactUserIds': contactUserIds}, fromJson: (j) => SOSContactsResponse.fromJson(j));

  Future<void> registerPushToken(String token) async {
    if (userId == null) return;
    try {
      await _postMutation<Map<String, dynamic>>('auth.registerPushToken', body: {'userId': userId, 'pushToken': token}, fromJson: (j) => j);
    } catch (_) {}
  }

  Future<GratitudeWallResponse> getGratitudeWall(String circleId, {int weeksBack = 0}) =>
      _getQuery('gratitudes.getWall', input: {'circleId': circleId, 'weeksBack': weeksBack}, fromJson: (j) => GratitudeWallResponse.fromJson(j));

  Future<void> shareGratitude({required List<String> circleIds, required String gratitudeText, required bool isAnonymous, String? displayName}) =>
      _postMutation<Map<String, dynamic>>('gratitudes.share', body: {'circleIds': circleIds, 'gratitudeText': gratitudeText, 'isAnonymous': isAnonymous, if (displayName != null) 'displayName': displayName}, fromJson: (j) => j);

  Future<void> deleteGratitude(String circleId, String gratitudeId) =>
      _postMutation<Map<String, dynamic>>('gratitudes.delete', body: {'circleId': circleId, 'gratitudeId': gratitudeId}, fromJson: (j) => j);

  Future<GratitudeNewCountResponse> getGratitudeNewCount(String circleId) =>
      _getQuery('gratitudes.getNewCount', input: {'circleId': circleId}, fromJson: (j) => GratitudeNewCountResponse.fromJson(j));

  Future<void> markGratitudesSeen(String circleId) =>
      _postMutation<Map<String, dynamic>>('gratitudes.markSeen', body: {'circleId': circleId}, fromJson: (j) => j);

  Future<GratitudeWeekCountResponse> getGratitudeWeekCount(String circleId) =>
      _getQuery('gratitudes.getWeekCount', input: {'circleId': circleId}, fromJson: (j) => GratitudeWeekCountResponse.fromJson(j));

  Future<CircleHeatmapResponse> getCircleHeatmap(String circleId, {int weekCount = 1}) =>
      _getQuery('circles.getHeatmap', input: {'circleId': circleId, 'weekCount': weekCount}, fromJson: (j) => CircleHeatmapResponse.fromJson(j));

  Future<CircleMilestonesResponse> getCircleMilestones(String circleId) =>
      _getQuery('circles.getMilestones', input: {'circleId': circleId}, fromJson: (j) => CircleMilestonesResponse.fromJson(j));

  Future<void> submitHeatmapData(String circleId, List<Map<String, dynamic>> weekData) =>
      _postMutation<Map<String, dynamic>>('circles.submitHeatmapData', body: {'circleId': circleId, 'weekData': weekData}, fromJson: (j) => j);

  // Private HTTP helpers

  Future<T> _getQuery<T>(String procedure, {Map<String, dynamic>? input, required T Function(dynamic) fromJson}) async {
    var urlString = '$_baseURL/api/trpc/$procedure';
    if (input != null) {
      final encoded = Uri.encodeComponent(jsonEncode(input));
      urlString += '?input=$encoded';
    }
    final uri = Uri.parse(urlString);
    final response = await http.get(uri, headers: await _buildAuthHeaders());
    _checkStatus(response);
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    return fromJson(body['result']['data'] as Map<String, dynamic>);
  }

  Future<T> _postMutation<T>(String procedure, {required Map<String, dynamic> body, required T Function(Map<String, dynamic>) fromJson}) async {
    final uri = Uri.parse('$_baseURL/api/trpc/$procedure');
    final authHeaders = await _buildAuthHeaders();
    final response = await http.post(uri, headers: {'Content-Type': 'application/json', ...authHeaders}, body: jsonEncode(body));
    _checkStatus(response);
    final responseBody = jsonDecode(response.body) as Map<String, dynamic>;
    return fromJson(responseBody['result']['data'] as Map<String, dynamic>);
  }

  Future<Map<String, String>> _buildAuthHeaders() async {
    final user = FirebaseAuth.instance.currentUser;
    if (user == null) return {};
    final token = await user.getIdToken();
    return {'Authorization': 'Bearer $token'};
  }

  void _checkStatus(http.Response response) {
    if (response.statusCode < 200 || response.statusCode > 299) {
      if (response.statusCode == 401) throw APIError.unauthorized;
      throw APIError.serverError;
    }
  }
}
